### PR TITLE
✨feat(rasa): automate the launch of Rasa stuff

### PR DESCRIPTION
**Pull request by RomanPlusPlus**
- add a script that launches Rasa stuff, with all the necessary preparations
- minor formatting fixes

### CODE REVIEW

File: ai_replica/utils/files.py [hunk 1]
```diff
@@ -2,9 +2,84 @@
...
+
+def get_main_dir_path(parent_level=1):
+    """Returns the path to the dir where all the replica code is located.
+
+    For example, if the path to this very module is '/stuff/AI-replica/ai_replica/utils/files.py',
+    then this func will return '/stuff/AI-replica/'
+
+    parent_level indicates how many levels one must go to reach the main dir,
+    counting from the level of the dir where this script is located.
+    Unless you want to move this module to some other place, there is no rason to change the parent_level.
```
Comment (ivan-karlovich) [inline ai_replica/utils/files.py]: Should it be "reason" instead of "rason"?

File: ai_replica/utils/system.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+def create_env_if_not_exist(env_path):
+    """Creates a venv virtual environment"""
+    if not os.path.exists(env_path):
+        subprocess.call(["python3", "-m", "venv", env_path])
+        print("created" + str(env_path))
+        existed7 = False
+    else:
+        print(str(env_path) + " already exist")
+        existed7 = True
+    return existed7
```
Comment (ivan-karlovich) [inline ai_replica/utils/system.py:42]: Was "7" in "existed7" added by purpose or is it a typo? Looks like it is indication of '?'?

File: ai_replica/utils/system.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+
+def install_piped_requirements(pip_path, requirements_path):
```
Comment (ivan-karlovich) [inline ai_replica/utils/system.py]: What are "piped requirements"?

File: rasa_launcher.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+"""
+
+import sys
+import time
+
+from ai_replica.utils.system import (
+    create_env_if_not_exist,
+    install_piped_requirements,
+    execute_command,
+    open_url_in_browser, is_expected_python_version,
```
Comment (ivan-karlovich) [inline rasa_launcher.py]: Should the last import be placed on a new line?

File: rasa_launcher.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+)
+from rasa.utils import (
+    get_rasa_paths,
+    start_rasa_actions_server,
+    start_rasa_main_server,
+    train_model,
+)
+
+required_python = "3.8"  # Rasa does not support later versions yet, as of Jan 2022
+expected_launch_duration_sec = 30  # # to avoid opening the browser-UI too early
```
Comment (ivan-karlovich) [inline rasa_launcher.py]: Not sure if the second "#" is needed here.

File: ai_replica/utils/system.py [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+    | "3"      | "3.8.1" | True   |
+    """
+    actual_version = platform.python_version()
+    return actual_version.startswith(expected_version)
+
+
+def create_env_if_not_exist(env_path):
+    """Creates a venv virtual environment"""
+    if not os.path.exists(env_path):
+        subprocess.call(["python3", "-m", "venv", env_path])
```
Comment (ivan-karlovich) [inline ai_replica/utils/system.py]: It looks like this script does not work as expected in the following case. The OS has python3.9 installed by default which is accessible via 'python3' command. In addition, python3.8. is installed and is accessible via 'python3.8' command. In this case ''python3 -m venv <path>" command creates python3.9 venv that leads to failure when installing Rasa. But "python3.8 -m venv <path>" works perfectly.

Probably, name of (or path to) python3.8 exec (or 3.7 or whatever version is required by Rasa and installed on the machine) can be provided as an additional parameter to "create_env_if_not_exist" function or set up as env variable which is consumed by create_env_if_not_exist. This env var can be set manually when rasa_launcher.py is launched.

**Review by ivan-karlovich — APPROVED** [review]
LGTM.

Just a couple of things:
- Minor issues/questions.
- Issue when python3.8 and python3.9 are simultaneously installed on the machine.
- When the script is being run for the second time, the "port is in use" issue appears. It can create inconvenience in case e.g. a Rasa action is updated or added and the actions server needs to be re-started for the changes to be propagated. Is it possible to start the server in a separate terminal window so that it can be re-started at all and independently from Rasa and bot servers? Independent restart of servers can speed up development process.

File: ai_replica/utils/files.py (hunk 1 above)
Comment (RomanPlusPlus) [reply ai_replica/utils/files.py]: Fixed

File: ai_replica/utils/system.py (hunk 2 above)
Comment (RomanPlusPlus) [reply ai_replica/utils/system.py:42]: You're right, it's "?". Using it for bools to make the code more readable (e.g. "existed?"). From my experience, it makes a difference in ML projects, where there are typically tons of bools.  

But i don't have hard feelings about it, can remove such stuff.

File: ai_replica/utils/system.py (hunk 3 above)
Comment (RomanPlusPlus) [reply ai_replica/utils/system.py]: Fixed: renamed it to `install_requirements_with_pip` for a better readability

File: rasa_launcher.py (hunk 4 above)
Comment (RomanPlusPlus) [reply rasa_launcher.py]: Fixed

File: rasa_launcher.py (hunk 5 above)
Comment (RomanPlusPlus) [reply rasa_launcher.py]: Fixed

File: ai_replica/utils/system.py (hunk 6 above)
Comment (RomanPlusPlus) [reply ai_replica/utils/system.py]: Fixed. For the env and the requirements installation, the script now uses the python version defined in config.yaml

**Comment by RomanPlusPlus** [comment]
>    When the script is being run for the second time, the "port is in use" issue appears. It can create inconvenience in case e.g. a Rasa action is updated or added and the actions server needs to be re-started for the changes to be propagated. Is it possible to start the server in a separate terminal window so that it can be re-started at all and independently from Rasa and bot servers? Independent restart of servers can speed up development process.

Fixed. The script now stops the servers before relaunching them. Also launches them in separate terminals.

Additionally, it's now possible to restart only selected servers by using command-line arguments of the rasa launcher.

File: ai_replica/utils/system.py (hunk 2 above)
Comment (ivan-karlovich) [reply ai_replica/utils/system.py:42]: It is OK for me to leave it. Just wanted to clarify.

**Pull request merged**